File: 面试题11-旋转数组的最小数字.py
class Solution:
    def minArray(self, numbers: List[int]) -> int:
        left = 0
        right = len(numbers) - 1

        while left < right:
            middle = (left + right) // 2
            if numbers[middle] < numbers[right]:
                right = middle        # 不排除middle就是最小值的可能性，right取middle
            elif numbers[middle] > numbers[right]:
                left = middle + 1     # middle比right大，不可能是最小值，left取middle+1
            else:
                right = right - 1
        
        return numbers[right]

File: 面试题11-旋转数组的最小数字.cpp
class Solution {
public:
    int minArray(vector<int>& numbers) {
        int left = 0, right = numbers.size() - 1, middle;

        while(left < right){
            middle = left + (right - left) / 2;
            if(numbers[middle] < numbers[right])
                right = middle;    // middle小，不能完全排除其为最小值的可能性
            else if(numbers[middle] > numbers[right])
                left = middle + 1;   // middle大，不可能为最小值
            else
                right = right - 1;
        }

        return numbers[right];
    }
};
